src/blockencodings.cpp

```diff
@@ -40,133 +65,31 @@ void CBlockHeaderAndShortTxIDs::FillShortTxIDSelector() const {
         if (lastprefilledindex > std::numeric_limits<uint16_t>::max())
-            return READ_STATUS_INVALID;
-        if ((uint32_t)lastprefilledindex > cmpctblock.shorttxids.size() + i) {
+            throw std::invalid_argument("tx index overflows");
+
+        if (static_cast<uint32_t>(lastprefilledindex) > cmpctblock.shorttxids.size() + i) {
             // If we are inserting a tx at an index greater than our full list of shorttxids
             // plus the number of prefilled txn we've inserted, then we have txn for which we
             // have neither a prefilled txn or a shorttxid!
-            return READ_STATUS_INVALID;
+            throw std::invalid_argument("invalid index for tx");
         }
-        txn_available[lastprefilledindex].reset(new CTransaction(cmpctblock.prefilledtxn[i].tx));
     }
-    prefilled_count = cmpctblock.prefilledtxn.size();
-
-    // Calculate map of txids -> positions and check mempool to see what we have (or dont)
-    // Because well-formed cmpctblock messages will have a (relatively) uniform distribution
-    // of short IDs, any highly-uneven distribution of elements can be safely treated as a
-    // READ_STATUS_FAILED.
-    std::unordered_map<uint64_t, uint16_t> shorttxids(cmpctblock.shorttxids.size());
-    uint16_t index_offset = 0;
-    for (size_t i = 0; i < cmpctblock.shorttxids.size(); i++) {
-        while (txn_available[i + index_offset])
-            index_offset++;
-        shorttxids[cmpctblock.shorttxids[i]] = i + index_offset;
-        // Bucket selection is a simple Binomial distribution. If we assume blocks of
-        // 10,000 transactions, allowing up to 12 elements per bucket should only fail
-        // once every ~1.3 million blocks and once every 74,000 blocks in a worst-case
-        // 16,000-transaction block.
-        if (shorttxids.bucket_size(shorttxids.bucket(cmpctblock.shorttxids[i])) > 12)
-            return READ_STATUS_FAILED;
-    }
-    // TODO: in the shortid-collision case, we should instead request both transactions
-    // which collided. Falling back to full-block-request here is overkill.
-    if (shorttxids.size() != cmpctblock.shorttxids.size())
-        return READ_STATUS_FAILED; // Short ID collision
-
-    std::vector<bool> have_txn(txn_available.size());
-    LOCK(pool->cs);
-    for (CTxMemPool::txiter it = pool->mapTx.begin(); it != pool->mapTx.end(); it++) {
-        std::unordered_map<uint64_t, uint16_t>::iterator idit = shorttxids.find(cmpctblock.GetShortID(it->GetTx().GetHash()));
-        if (idit != shorttxids.end()) {
-            if (!have_txn[idit->second]) {
-                txn_available[idit->second].reset(new CTransaction(it->GetTx()));
-                have_txn[idit->second]  = true;
-                mempool_count++;
-            } else {
-                // If we find two mempool txn that match the short id, just request it.
```

Comment:
Removing the collision handling cases (unless I missed where you added them elseswhere) will dramatically reduce the effectiveness of reconstruction with large mempools. The code here was driven by doing the math, and by testing it, not some lark. :) just thought you should know.